packages/core/strapi/lib/core/loaders/apis.js

```diff
@@ -1,20 +1,159 @@
-  const apiDir = join(strapi.dir, 'api');
+  const apisDir = join(strapi.dir, 'api');
 
-  if (!existsSync(apiDir)) {
+  if (!existsSync(apisDir)) {
     throw new Error(`Missing api folder. Please create one in your app root directory`);
   }
 
-  const apis = await loadFiles(apiDir, '*/!(config)/**/*.*(js|json)');
-  const apiConfigs = await loadConfig(apiDir, '*/config/**/*.*(js|json)');
+  const apisFDs = await fse.readdir(apisDir, { withFileTypes: true });
+  const apis = {};
+
+  // only load folders
+  for (const apiFD of apisFDs) {
+    if (apiFD.isDirectory()) {
+      const apiName = normalizeName(apiFD.name);
+      const api = await loadAPI(join(apisDir, apiFD.name));
+
+      apis[apiName] = api;
+    }
+  }
+
+  validateContentTypesUnicity(apis);
+
+  for (const apiName in apis) {
+    strapi.container.get('apis').add(apiName, apis[apiName]);
+  }
+};
+
+const validateContentTypesUnicity = apis => {
+  const allApisSchemas = Object.values(apis).flatMap(api => Object.values(api.contentTypes));
+
+  const names = [];
+  allApisSchemas.forEach(({ schema }) => {
+    if (schema.info.singularName) {
+      const singularName = _.kebabCase(schema.info.singularName);
+      if (names.includes(singularName)) {
+        throw new Error(`The singular name "${schema.info.singularName}" should be unique`);
+      }
+      names.push(singularName);
+    }
+
+    if (schema.info.pluralName) {
+      const pluralName = _.kebabCase(schema.info.pluralName);
+      if (names.includes(pluralName)) {
+        throw new Error(`The plural name "${schema.info.pluralName}" should be unique`);
+      }
+      names.push(pluralName);
+    }
+  });
+};
+
+const loadAPI = async dir => {
+  const [
+    index,
+    config,
+    routes,
+    controllers,
+    services,
+    policies,
+    middlewares,
+    contentTypes,
+  ] = await Promise.all([
+    loadIndex(dir),
+    loadDir(join(dir, 'config')),
+    loadDir(join(dir, 'routes')),
+    loadDir(join(dir, 'controllers')),
+    loadDir(join(dir, 'services')),
+    loadDir(join(dir, 'policies')),
+    loadDir(join(dir, 'middlewares')),
+    loadContentTypes(join(dir, 'content-types')),
+  ]);
+
+  return {
+    ...(index || {}),
+    config: config || {},
+    routes: routes || [],
+    controllers: controllers || {},
+    services: services || {},
+    policies: policies || {},
+    middlewares: middlewares || {},
+    contentTypes: contentTypes || {},
+  };
+};
+
+const loadIndex = async dir => {
+  if (await fse.pathExists(join(dir, 'index.js'))) {
+    return loadFile(join(dir, 'index.js'));
+  }
+};
 
-  return _.merge(apis, apiConfigs);
+const loadContentTypes = async dir => {
+  if (!(await fse.pathExists(dir))) {
+    return;
+  }
+
+  const fds = await fse.readdir(dir, { withFileTypes: true });
+  const contentTypes = {};
+
+  // only load folders
+  for (const fd of fds) {
+    if (fd.isFile()) {
+      continue;
+    }
+
+    const contentTypeName = normalizeName(fd.name);
```

Comment:
You may want to use `kebabCase` instead. Otherwise the singularName and the key of the ct may be different and it would throw at the validation step.